src/dev_wasi_nn.erl

```diff
@@ -0,0 +1,324 @@
+        % Use persistent context management (fast if model already loaded)
+    case dev_wasi_nn_nif:switch_model(ModelPath, ModelConfig) of
+        {ok, Context} ->
+            % Create or reuse session-specific execution context
+            case dev_wasi_nn_nif:init_execution_context_once(Context, SessionId) of
+                {ok, ExecContextId} ->
+                    % Run inference with session-specific context
+                    case dev_wasi_nn_nif:run_inference(Context, ExecContextId, binary_to_list(Prompt)) of
+                        {ok, Output} ->
+                            ?event(output, Output),
+                            {ok, #{
+                                <<"result">> => Output,
+                                <<"session-id">> => list_to_binary(SessionId)
+                            }};
+                        {error, Reason} ->
+                            ?event(dev_wasi_nn, {inference_failed, SessionId, Reason}),
+                            {error, Reason}
+                    end;
+                {error, Reason2} ->
+                    ?event(dev_wasi_nn, {session_init_failed, SessionId, Reason2}),
+                    {error, Reason2}
+            end;
+        {error, Reason3} ->
+            ?event(dev_wasi_nn, {model_load_failed, SessionId, ModelPath, Reason3}),
+            {error, Reason3}
+    end.
+
+%% @doc Download and retrieve a model by Arweave transaction ID.
+%% This function handles the complete model retrieval workflow including:
+%% - Starting the HTTP server for Arweave gateway access
+%% - Configuring local filesystem caching to avoid repeated downloads
+%% - Downloading the model from Arweave if not already cached
+%% - Resolving the local file path where the model is stored
+%%
+%% The function uses a two-tier storage strategy:
+%% 1. First checks local cache (hb_store_fs) for existing model
+%% 2. Falls back to Arweave gateway (hb_store_gateway) if not cached
+%% 3. Automatically caches downloaded models locally for future use
+%%
+%% @param TxID The Arweave transaction ID containing the model file as a binary.
+%% @returns {ok, LocalFilePath} where LocalFilePath is a string path to the 
+%%          cached model file, or {error, Reason} on failure.
+read_model_by_ID(TxID) ->
+    %% Start the HTTP server (required for gateway access)
+    hb_http_server:start_node(#{}),
+    %% Configure store with local caching for model files
+    LocalStore = #{
+        <<"store-module">> => hb_store_fs,
+        <<"name">> => <<"model-cache">>
+    },
+    Opts = #{
+        store => [
+            %% Try local cache first
+            LocalStore,
+            #{
+                <<"store-module">> => hb_store_gateway,
+                %% Cache results here
+                <<"local-store">> => LocalStore
+            }
+        ]
+    },
+    %% Attempt to read the model from cache or download from Arweave
+    case hb_cache:read(TxID, Opts) of
```

Comment:
Best practice is just to pass the `Opts` to your helper function, then you can do `hb_cache:read(TXID, Opts)`. I guess the idea here was to add the specific FS model store to the opts? In which case, you are better off making a function like this:

```
opts(BaseOpts) ->
    ModelStore = ..., % Ideally checking the opts for a `model_store` param so that the user can configure it if they want.
    NewOpts#{
        store = [ModelStore|hb_opts:get(store, [], BaseOpts)]
    }.
```
If the concern is just the size of the models though, it might make more sense to just work with us on the relevance filter for the `hb_store` interface. We want to add the ability to have filters based on size or path prefix (e.g, `data/`) to `hb_store:write`, such that bigger items go to the FS, etc.